base/prometheus/prometheus.ClusterRoleBinding.yaml

```diff
@@ -15,4 +15,4 @@ subjects:
 - kind: ServiceAccount
   name: prometheus
   # value of namespace needs to match namespace value in base/prometheus/prometheus.ServiceAccount.yaml
-  namespace: default
+  namespace: ns-sourcegraph
```

Comment:
Can you explain why this gets changed from the `default` namespace to `ns-sourcegraph` when the rest of sourcegraph is still set to deploy into the `default` namespace? I know **if** a customer uses kustomize it gets changed, but wouldn't it be better to change this namespace with kustomize as well, instead of having the namespaces diverge like this?

Reply:
this namespace is the default namespace used in the repo currently: https://github.com/sourcegraph/deploy-sourcegraph/pull/1578#discussion_r553379376

Reply:
Only for some overlays...not all of them. For example non-root overlays don't apply a different ns. If you clone this repo, and apply the base as is, everything else including the prometheus deploymet is deployed into default. 

My suggestion would be something like this:
https://github.com/sourcegraph/deploy-sourcegraph/commit/6d2b115eb7e5402662e8e0bfa7105ca8ec8c28a1

Reply:
Ah I see! Completely misunderstood this, thanks!

Looking at your patch, does this mean we have to ensure both the base and overlay are kept in-sync manually?

Reply:
I looked into kustomize and I can't see a way to edit one line in otherwise equal configmap, so unfortunately, yes we'd need to keep this in sync "manually" but the only change it the namespace on that one line. I don't see why we couldn't just do a find/replace on that line in some automated fashion

Reply:
I think that's a sufficient solution as we hold out for Dhall - do you mind pushing your patch to this branch?

also cc @pecigonzalo for a look again

Reply:
@davejrt I cherry-picked your change here: [`93e5188` (#1578)](https://github.com/sourcegraph/deploy-sourcegraph/pull/1578/commits/93e5188d660bfd66bf8de56cc7004c5b07e7c104)

Reply:
Just looking at this in its uncommented format, wouldn't this break monitoring for anyone deploying into an implicit namespace (`kubectl ns foo && kubectl apply` )?

Reply:
I think it's better off actually deleting this altogether, as we don't explicitly specify a namespace anywhere else in our manifests. If someone uses kustomize to change the namespace, these clusterrole bindings will be changed. If they apply to an alternate namespace implicitly as per @pecigonzalo example, they aren't affected.

Reply:
> I think it's better off actually deleting this altogether

this configuration is pretty important to have in most environments, and this PR revolves around adding this bit - see description.

I'm curious as to why we don't deploy to a namespace by default? Seems ideal to be able to identify Sourcegraph services separately from whatever else is running in a customer's environment by default

Reply:
I did some digging, and the following Kubernetes metadata is the only thing we can get on a per-container label level (which seems to be the only level at which is exported by cadvisor):

- `io.kubernetes.container.name`
- `io.kubernetes.pod.name`
- `io.kubernetes.pod.namespace`
- `io.kubernetes.pod.uid`

`name` alone is not sufficient, as described

Reply:
an alternative solution is that cAdvisor can be configured to export certain container env variables as labels. We can then require that all containers that want to be monitored should set the following env:

```yml
        env:
        - name: CADVISOR_SHOULD_MONITOR
          value: true
```

Reply:
removed the namespace declarations in 1aa44df8f2100967885eea183bc77f5a91821f49

Reply:
> I'm curious as to why we don't deploy to a namespace by default? Seems ideal to be able to identify Sourcegraph services separately from whatever else is running in a customer's environment by default

We do want customers to deploy in a namespace, but the name and such is generally defined by the end-user in most cases as they might have some naming conventions like `ns-<name>` or `prod-<name` or a combination of those or others.
If we add the namespace to all resources, the customer would have to either create the same name on its own, or we will have to bundle a Namespace resoruces. In both cases, they will have to modify all files in many places to make them match if they dont use the bundled one.

Kustomize/Dhall resolve this last part AFAIK but not the templating in the config file. Maybe the pod can template the file on startup with an init container or process using the `KUBE_NAMESPACE` env var.

Reply:
Do we want to explore a long-term solution to this? I think I've already sunk a huge amount of time trawling forums and docs and even source code for both the original cAdvisor issues and ergonomic ways to try and resolve this filtering problem

I think there are a lot of solutions here including something like:

> Maybe the pod can template the file on startup with an init container or process using the KUBE_NAMESPACE env var.

I have some other ideas as well, but they all seem pretty hack and could probably do with a more detailed comparison of how each method will work out _and_ how they might map to dhall. I've also been pondering whether we want to stick with cAdvisor at all, maybe there are more modern collectors out there with the versatility we need?

I'm tempted right now to just revert this PR to the first commit (f886f2fab31c3b313833e9c94db956061f59a55c) and merge that, and let customers know how to add this config if they encounter issues

Reply:
I think that sounds good and lets just open or keep the original issue for this to track those other ideas, so we dont start from scratch.

Reply:
Opened https://github.com/sourcegraph/deploy-sourcegraph/pull/1644 , will open an issue for this stuff

Reply:
Follow-up: https://github.com/sourcegraph/sourcegraph/issues/17365